synapse/storage/databases/state/store.py

```diff
@@ -228,6 +251,142 @@ def _get_state_for_group_using_cache(
 
         return state_filter.filter_state(state_dict_ids), not missing_types
 
+    def _get_state_for_group_gather_inflight_requests(
+        self, group: int, state_filter: StateFilter
+    ) -> Tuple[Sequence[AbstractObservableDeferred[StateMap[str]]], StateFilter]:
+        """
+        Attempts to gather in-flight requests and re-use them to retrieve state
+        for the given state group, filtered with the given state filter.
+
+        Returns:
+            Tuple of two values:
+                A sequence of ObservableDeferreds to observe
+                A StateFilter representing what else needs to be requested to fulfill the request
+        """
+
+        inflight_requests = self._state_group_inflight_requests.get(group)
+        if inflight_requests is None:
+            # no requests for this group, need to retrieve it all ourselves
+            return (), state_filter
+
+        state_filter_left_over = state_filter
+        reusable_requests = []
+        for (
+            request_state_filter,
+            request_deferred,
+        ) in inflight_requests.items():
+            new_state_filter_left_over = state_filter_left_over.approx_difference(
+                request_state_filter
+            )
+            if new_state_filter_left_over != state_filter_left_over:
+                # reusing this request narrows our StateFilter down a bit.
+                reusable_requests.append(request_deferred)
+            state_filter_left_over = new_state_filter_left_over
```

Comment:
I'm finding some of this a bit confusing as there are three variables that are similar (`state_filter`, `new_state_filter_left_over`, and `state_filter_left_over`). I think one of those is unneeded and we could just rename the input param to `state_filter_left_over`.

I also don't this line and the below can only occur if `new_state_filter_left_over != state_filter_left_over`? But I'm not 100% sure. It might be clearer to check if they're equal and then continue?

Reply:
I've switched it around following your suggestions